### PR TITLE
Prevent users in invalid state

### PR DESCRIPTION
**Pull request by betesh**

### CODE REVIEW

File: lib/plaid/user.rb [hunk 1]
```diff
@@ -332,11 +334,17 @@ def delete
     # Returns another User record with the same access token, but tied to the
     # new product.
     def upgrade(product)
+      User.upgrade product, access_token, client: client
+    end
+
+    # Internal: Upgrade a User instance.
+    def self.upgrade(product, access_token, client: nil)
```
Comment (be9) [inline lib/plaid/user.rb]: Why add this method? It's not used anywhere else.

Comment (betesh) [inline lib/plaid/user.rb]: It's used in `User#upgrade` (the instance method), 4 lines up.

`User#upgrade` used to call User.new directly, which is now private.  With this additional layer, while it's still possible to bypass `User#upgrade` and call `User.upgrade` (the class method), the resulting User would be in a valid state, as opposed to if it were possible to call `User.new`, which would result in a User in an invalid state.

Comment (be9) [inline lib/plaid/user.rb]: I don't like this method. It's sole purpose is to workaround the `private_class_method :new` which you introduce in the same commit.

I think checking the product in the constructor would be enough. And `private_class_method :new` isn't needed at all.

**Comment by be9** [comment]
Closing this for now.

**Pull request closed without merging**